### ISSUE TITLE
FMU Cross-compilation on GRT target

### ISSUE THREAD

**Issue by FranciscoCollado** [CATIA-Systems/FMIKit-Simulink#79]
How can we do a cross-compilation of a FMU model, from Windows to Linux, using grtfmi.tlc on a windows PC?

**Comment by t-sommer** [comment]
Since this [seems to be supported by CMake](https://gitlab.kitware.com/cmake/community/wikis/doc/cmake/CrossCompiling) it should also work for grtfmi.tlc. Can you describe a use case where you want to use this feature?

**Comment by chrbertsch** [comment]
One use case from our side is that  FMUs are exported from a Windows PC  but shall be used on several platforms (windows, linux). So it would be great to be able to select the needed binaiy platforms and create them seamlessly

**Comment by FranciscoCollado** [comment]
My desktop environment is  a Windows PC which I use it for creating FMU models and run them offline, on my desktop PC.
 For HIL, where we need to run them online on realtime, the system is using a Linux operative system, so the requirement of the simulation software installed on the HIL system is to have the FMU compiled for Linux (.so file instead of .dll) in order to be able to run it.

**Comment by t-sommer** [comment]
Usually it's easier to export a source code FMU and compile the binary on the target system. Would this be an option for you?

**Comment by chrbertsch** [comment]
> Usually it's easier to export a source code FMU and compile the binary on the target system.

Often the person who creates the FMUs is more experienced with FMI (and possibly installing compilers etc. ) than the person who consumes it. So it would be very convenient to add multiple targets at generation time ...

Do I remember right, that with the GUI version of fmpy you can add a binary for the _current_  platform to the FMU? 

I think the intent of this issue is the same as addressed by
https://sweng.au.dk/fmubuilder
https://github.com/INTO-CPS-Association/fmu-builder
@CThuleHansen: could you please comment on the status of fmu-builder  and if it will be further maintained? If one would not be able to upload FMUs to fmubuilder due to IP protection: What would one have to do to set up one's own fmubuilder-server=

**Comment by FranciscoCollado** [comment]
As Christian says, it would be easier for us to be able to compile the FMU into multiple platform targets at the generation time. As far as I know another FMU tools will create automatically the libraries for Win/Linux/Mac automatically without specifying the platform.

**Comment by t-sommer** [comment]
> As far as I know another FMU tools will create automatically the libraries for Win/Linux/Mac automatically without specifying the platform.

Which tool can do that?

The only feasible option to compile a Linux binary on Windows is IMHO a Linux container (Docker/Linux subsystem for Windows).

**Comment by FranciscoCollado** [comment]
Mathworks FMU tool export:
https://uk.mathworks.com/matlabcentral/fileexchange/68814-tool-coupling-co-simulation-fmu-export-for-simulink

**Comment by t-sommer** [comment]
The "Tool-Coupling Co-Simulation FMU Export for Simulink" is a generic wrapper that connects to a  Simulink instance. It comes pre-compiled for all platforms.

**Comment by CThuleHansen** [comment]
Hi,

@chrbertsch the future of fmu builder is a bit uncertain, as our whole server infrastructure has changed, and we are currently busy with another project.
I will discuss it with my team next week, but I expect that it **WILL be further maintained**. It is available at https://sweng.au.dk/fmubuilder

**Comment by StrAma84** [comment]
> Usually it's easier to export a source code FMU and compile the binary on the target system. Would this be an option for you?

Actually this would be an option for me.
I have a virtual machine running Linux/Ubuntu on a Windows host. gcc compiler and CMake are installed on the VM as well.
Originally I created the FMU in Windows.

Does anyone have experience how to rebuilt the binary in the Linux environment?
Any hint is welcome...

**Comment by chrbertsch** [comment]
> Does anyone have experience how to rebuilt the binary in the Linux environment?

I recommend to use fmpy. Works from the command line (https://github.com/CATIA-Systems/FMPy/blob/216bd08c76c05a97d145e29f8ada7a156df0bceb/fmpy/util.py#L824) and the gui (Help --> Create platform binary)

**Comment by StrAma84** [comment]
Thanks. In general that seems to be helpful.
However the option 'Compile Platform Binary' is greyed out (i.e. not available) in the gui, nor does it seem to be a valid fmpy command in the terminal.

I installed via conda.
Is there any package I missed to install?

**Comment by Tulgs** [comment]
Hello,

> Since this [seems to be supported by CMake](https://gitlab.kitware.com/cmake/community/wikis/doc/cmake/CrossCompiling) it should also work for grtfmi.tlc. Can you describe a use case where you want to use this feature?

do you have an idea how I could perform that? I tried to compile for Linux64 by including a custom toolchain in "grtfmi_make_rtw_hook.m" in the CMake call, but it still compiles as DLL, just with .so ending, so doesn't work on Linux but on Windows.

> I recommend to use fmpy. Works from the command line (https://github.com/CATIA-Systems/FMPy/blob/216bd08c76c05a97d145e29f8ada7a156df0bceb/fmpy/util.py#L824) and the gui (Help --> Create platform binary)

This option is greyed out for me either like for StrAma84 when I import the fmu (compiled in widows) in linux as well as in windows. Would be a nice solution however.

**Comment by chrbertsch** [comment]
> This option is greyed out for me either like for StrAma84 when I import the fmu (compiled in widows) in linux as well as in windows. Would be a nice solution however.

Could it be that you have binary-only FMUs w/o source code? Then it is greyed out for me as well.  Please check the sources folder of your FMU. And try with one of the source code FMUs with source code from the FMU cross check, e.g. https://github.com/modelica/fmi-cross-check/tree/master/fmus/2.0/cs/c-code/Dymola/2019FD01

**Comment by Tulgs** [comment]
> Could it be that you have binary-only FMUs w/o source code? Then it is greyed out for me as well. Please check the sources folder of your FMU. And try with one of the source code FMUs with source code from the FMU cross check, e.g. https://github.com/modelica/fmi-cross-check/tree/master/fmus/2.0/cs/c-code/Dymola/2019FD01

Yes they are binary-only. I'm compiling them in Simulink with grtfmi.tlc and do get the .fmu file with subfolder binaries only. How would I get the sources in it? Thanks for your fast response.

**Comment by chrbertsch** [comment]
> How would I get the sources in it? Thanks for your fast response.

read the documentation :-) https://github.com/CATIA-Systems/FMIKit-Simulink/blob/master/docs/fmu_export.md

**Comment by Tulgs** [comment]
> read the documentation :-) https://github.com/CATIA-Systems/FMIKit-Simulink/blob/master/docs/fmu_export.md

Thank you very much :)

**Comment by StrAma84** [comment]
Many thanks. I also missed to add the source code to the FMU.
However when I try to build the binary now in the Linux environment, I'm receiving the following error:
"(...) _private.h:100:2: error: #error Code was generated for compiler with different sized ulong/long. (...)"

Presumably it's related to certain Code Generation settings in Simulink. I tried several different settings, but yet not sucsesful.
Is there a common way to solve this?

**Comment by fajansen** [comment]
To solve this you have to go to the "Hardware Implementation" settings and change the Device vendor to "Custom Processor". Then you can change the number of bits for "long".

**Comment by StrAma84** [comment]
Thanks! That solved it.

**Comment by rrzhu** [comment]
@t-sommer 
When I use grtfmi_compile_wsl, there is an error grtfmi_wslpath
It looks undefined
